mesonbuild/environment.py

```diff
@@ -682,27 +700,56 @@ def move_compiler_options(properties: Properties, compopts: T.Dict[str, T.Defaul
         # unset, set those now.
 
         for for_machine in MachineChoice:
-            p_env_pair = get_env_var_pair(for_machine, self.coredata.is_cross_build(), 'PKG_CONFIG_PATH')
+            for key, envkey in OTHER_OPTS_TO_ENV.items():
+                p_env_pair = get_env_var_pair(for_machine, self.coredata.is_cross_build(), envkey)
+                if p_env_pair is not None:
+                    p_env_var, p_env = p_env_pair
+
+                    # these may contain duplicates, which must be removed, else
+                    # a duplicates-in-array-option warning arises.
+                    p_list = list(mesonlib.OrderedSet(re.split(r':|;', p_env)))
+
+                    if self.first_invocation:
+                        # Environment variables override config
+                        self.meson_options[for_machine][''][key] = p_list
+                    elif self.meson_options[for_machine][''].get(key, []) != p_list:
+                        mlog.warning(
+                            p_env_var,
+                            'environment variable does not match configured',
+                            'between configurations, meson ignores this.',
+                            'Use -D{} instead'.format(key)
+                        )
+
+            compiler_args_from_env = False
+            for key, envkey in COMPILER_OPTS_TO_ENV.items():
+                p_env_pair = get_env_var_pair(for_machine, self.coredata.is_cross_build(), envkey)
+                if p_env_pair is not None:
+                    p_env = p_env_pair[1]
+
+                    if self.first_invocation:
+                        # Environment variables override config
+                        self.compiler_options[for_machine][key]['args'] = split_args(p_env)
+                        compiler_args_from_env = True
+
+            # Handle CPPFLAGS, which are only relavent for a couple of languages
+            p_env_pair = get_env_var_pair(for_machine, self.coredata.is_cross_build(), 'CPPFLAGS')
+            if p_env_pair is not None:
+                p_env = split_args(p_env_pair[1])
+                for l in ['c', 'cpp', 'objc', 'objcpp']:
+                    compargs = self.compiler_options[for_machine][l]
+                    # If the compiler args were set from the environment extend
+                    # them with CPPFLAGS, if they came from somewhere else
+                    # replace them.
+                    if compiler_args_from_env and 'args' in compargs:
```

Comment:
Shouldn't all this be done only during the `self.first_invocation` or am I missing something?

Reply:
The more I think about it, the more I want to wrap this entire block in `if self.first_invocation`, hmm.

Reply:
fwiw, I do have plans to remove all those `self.first_invocation`, meson can reevaluate options and pick new defaults since we have `meson-private/cmd_line.txt` that tells what are the options given by the user. This is a recurring issue people constantly raise, changing default_options has no effect until you --wipe.